rholang/src/main/scala/coop/rchain/rholang/interpreter/compiler/Compiler.scala

```diff
@@ -12,36 +12,50 @@ import coop.rchain.rholang.interpreter.errors._
 import coop.rchain.rholang.syntax.rholang_mercury.Absyn.Proc
 import coop.rchain.rholang.syntax.rholang_mercury.{parser, Yylex}
 
-trait ParBuilder[F[_]] {
-  def buildNormalizedTerm(source: String, normalizerEnv: Map[String, Par]): F[Par]
+trait Compiler[F[_]] {
 
-  def buildNormalizedTerm(reader: Reader, normalizerEnv: Map[String, Par]): F[Par]
+  def sourceToADT(source: String): F[Par] =
+    sourceToADT(source, Map.empty[String, Par])
```

Comment:
I just want to say that this overloaded method can be a syntax extension.

Syntax extension is brought by @tgrospic which can make trait simpler. Here is the example.
https://github.com/rchain/rchain/blob/dev/rspace/src/main/scala/coop/rchain/rspace/state/RSpaceExporterSyntax.scala

This is just a good to have request.

Reply:
@tgrospic Did you find this pattern somewhere, or is it your own?

Reply:
It looks like the "Pimp My Class" pattern: http://blog.rcard.in/jvm/programming/design-pattern/2019/12/15/pimp-my-library-pattern.html

Reply:
Still, that's not appropriate here. We have control of the `Compiler` interface. Moreover, each of these overloaded alternatives are common enough to merit adding them to the main API.

Reply:
@JosDenmark this is the pattern used in cats. Here is the example for monad.
https://github.com/typelevel/cats/blob/e1a7cfcddc/core/src/main/scala/cats/syntax/monad.scala
And all extension can then be imported with one import _all_.
https://github.com/typelevel/cats/blob/e1a7cfcddc/core/src/main/scala/cats/syntax/all.scala

I like it because it gives more flexibility to define extensions only in projects that are needed or to combine with additional dependency to enrich the _syntax_.
My opinion is that default implementations are fine if they don't require additional implicit dependencies like here.
https://github.com/rchain/rchain/blob/b69eb8a457f67e7ca32581c515db6958515cd62c/block-storage/src/main/scala/coop/rchain/blockstorage/BlockStore.scala#L21-L22

So here I'm fine with either decision. I'm thinking more about how to make it more clear which methods are basic definitions and which are defaults/extensions.

P.S.
This is basically dependency injection pattern which will be used even more in v3 of cats. There are also changes to the inheritance hierarchy where `Sync` will be moved from top to bottom.
https://youtu.be/xEkSDiz3XO0?t=2449

Reply:
Yeah. I don't think these qualify as "extension" methods. They belong in the main interface IMO. Let's consider this issue resolved.